src/datasets/utils/remote_utils.py

```diff
@@ -0,0 +1,177 @@
+            callback (callable, optional): Callback function to be called with each response data chunk.
+
+        Returns:
+            generator or type returned by `callback`
+        """
+        headers = copy.deepcopy(headers) or {}
+        headers["user-agent"] = cls.get_datasets_user_agent(user_agent=headers.get("user-agent"))
+        if resume_size > 0:
+            headers["Range"] = "bytes=%d-" % (resume_size,)
+        response = cls.request_with_retry(
+            verb="GET",
+            url=url,
+            stream=True,
+            proxies=proxies,
+            headers=headers,
+            cookies=cookies,
+            max_retries=max_retries,
+        )
+        if response.status_code == 416:  # Range not satisfiable
+            return
+        if callback:
+            content_length = response.headers.get("Content-Length")
+            total = resume_size + int(content_length) if content_length is not None else None
+            not_verbose = bool(logger.getEffectiveLevel() > WARNING)
+            progress = tqdm(
+                unit="B",
+                unit_scale=True,
+                total=total,
+                initial=resume_size,
+                desc="Downloading",
+                disable=not_verbose,
+            )
+            for chunk in response.iter_content(chunk_size=1024):
+                if chunk:  # filter out keep-alive new chunks
+                    progress.update(len(chunk))
+                    callback(chunk)
+            progress.close()
+        else:
+            return response.iter_content(chunk_size=1024)
+
+    @staticmethod
+    def get_datasets_user_agent(user_agent: Optional[Union[str, dict]] = None) -> str:
+        ua = "datasets/{}; python/{}".format(__version__, config.PY_VERSION)
+        ua += "; pyarrow/{}".format(config.PYARROW_VERSION)
+        if config.TORCH_AVAILABLE:
+            ua += "; torch/{}".format(config.TORCH_VERSION)
+        if config.TF_AVAILABLE:
+            ua += "; tensorflow/{}".format(config.TF_VERSION)
+        if config.BEAM_AVAILABLE:
+            ua += "; apache_beam/{}".format(config.BEAM_VERSION)
+        if isinstance(user_agent, dict):
+            ua += "; " + "; ".join("{}/{}".format(k, v) for k, v in user_agent.items())
+        elif isinstance(user_agent, str):
+            ua += "; " + user_agent
+        return ua
+
+
+class FtpClient:
+    @staticmethod
+    def head(url, timeout=2.0):
+        try:
+            with urllib.request.urlopen(url, timeout=timeout) as r:
+                r.read(1)
+        except Exception:
+            return False
+        return True
+
+    @staticmethod
+    def get(url, timeout=2.0, callback=None):
+        try:
+            logger.info(f"Getting through FTP {url}")  # into {temp_file.name}")
+            with urllib.request.urlopen(url, timeout=timeout) as r:
+                if callback:
+                    callback(r)
```

Comment:
same here

Reply:
For this case, however, I need the file open while copying it by chunks, thus within the context manager scope. ;)

Reply:
I would remove the callback logic here too.
It's ok if the ftp get must be used with `with` to properly close the file.